Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -379,22 +379,58 @@ public virtual OrderEvent MarketOnOpenFill(Security asset, MarketOnOpenOrder ord
             // have large gaps, in which case the currentBar.EndTime will be in the past
             // ASUR  | | |      [order]        | | | | | | |
             //  SPY  | | | | | | | | | | | | | | | | | | | |
-            var currentBar = asset.GetLastData();
+            var openTime = DateTime.MinValue;
+            var endTime = DateTime.MinValue;
+
+            var subscribedTypes = GetSubscribedTypes(asset);
+
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                var trade = asset.Cache.GetAll<Tick>().LastOrDefault(x => x.TickType == TickType.Trade && x.Price > 0);
+                if (trade != null)
+                {
+                    openTime = trade.Time;
+                    endTime = trade.EndTime;
+                    fill.FillPrice = trade.Price;
+                }
+            }
+            else if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+                if (tradeBar != null)
+                {
+                    openTime = tradeBar.Time;
+                    endTime = tradeBar.EndTime;
+                    fill.FillPrice = tradeBar.Open;
+                }
+            }
+
             var localOrderTime = order.Time.ConvertFromUtc(asset.Exchange.TimeZone);
-            if (currentBar == null || localOrderTime >= currentBar.EndTime) return fill;
+            if (localOrderTime >= endTime) return fill;
 
             // if the MOO was submitted during market the previous day, wait for a day to turn over
             if (asset.Exchange.DateTimeIsOpen(localOrderTime) && localOrderTime.Date == asset.LocalTime.Date)
             {
                 return fill;
             }
 
+            // If the first data after the market opens is a quote bar (high resolution only),
+            // the open of the last trade bar is not the open for the current day, but the last open in cache.
+            // We need to verify whether the trade data is from the open market.
+            var resolution = Parameters.ConfigProvider.GetSubscriptionDataConfigs(asset.Symbol).GetHighestResolution();
+            if (resolution < Resolution.Hour && !asset.Exchange.DateTimeIsOpen(openTime))
```

Comment:
Suggest we move this change upwards into each tick or tradeBar `if` case, don't think we need to rely on the `.GetHighestResolution()`, we don't know the subscriptions it's looking at. For the `tick` case we can directly check `asset.Exchange.DateTimeIsOpen()`. For the tradeBar case we could call `tb.Period.ToHigherResolutionEquivalent()` and get the exact resolution of the bar we are going to use, think it would be a more transparent decision/logic